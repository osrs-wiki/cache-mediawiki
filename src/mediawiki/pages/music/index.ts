export { default } from "./music";
export * from "./music";