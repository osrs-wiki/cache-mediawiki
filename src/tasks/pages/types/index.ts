export * from "./area";
export * from "./item";
export * from "./music";
export * from "./npc";
export * from "./scenery";
