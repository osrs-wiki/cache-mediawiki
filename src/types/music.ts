/**
 * Represents a music track from DB table 44
 */
export interface MusicTrack {
  /** The internal sort name of the track */
  sortName: string;
  /** The display name of the track */
  displayName: string;
  /** Hint text for unlocking the track */
  unlockHint: string;
  /** Duration of the track in seconds */
  duration: number;
  /** The unique ID of the track */
  id: number;
}