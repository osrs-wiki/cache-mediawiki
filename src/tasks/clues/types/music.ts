import { cluePageBuilder } from "../../../mediawiki/pages/clue";
import {
  DBTable,
  DiskCacheProvider,
  FlatCacheProvider,
  Item,
  ParamID,
} from "../../../utils/cache2";
import { ITEM_PARAM_ID, getAnswer, getTier, writeClueFile } from "../utils";

const generateMusicPages = async (
  cache: Promise<FlatCacheProvider | DiskCacheProvider>
) => {
  const rows = await DBTable.loadRows(cache, 13);

  const items = await Item.all(cache);
  const paramMap: { [key: string | number]: Item } = {};
  items.forEach((item) => {
    if (
      item.name.includes("Clue scroll") &&
      item.params.has(ITEM_PARAM_ID as ParamID)
    ) {
      paramMap[item.params.get(ITEM_PARAM_ID as ParamID) as number] = item;
    }
  });

  rows.forEach(async (row) => {
    try {
      if (row) {
        const dbRowId = row.id;
        const values = row.values;
        const id = values[0][0] as string;
        const tier = getTier(values[1][0]);

        const answers = await getAnswer(cache, values[4]);
        const musicAnswer = await getAnswer(cache, values[2]);
        const songName = musicAnswer.map((answer) => answer.answer).join(", ");
        answers.push({
          answer: ` Play the [[music]] track [[${songName}]].`,
          entityName: "",
          type: "music",
          worldLocs: [],
        });

        const itemName = `Clue scroll (${tier})`;
        const item = paramMap[dbRowId as number];
        const builder = cluePageBuilder({
          id,
          tier,
          clue: `I'd like to hear some music.\nCome and see me on the bridge in\nFalador Park, and play:\n${songName}\nThanks,\nCecilia`,
          answers: answers,
          itemName,
          item,
          type: "music",
        });

        writeClueFile("music", itemName, songName, builder);
      }
    } catch (e) {
      console.error(`Error creating clue ${row.id}: ${e}`);
    }
  });

  console.log(`Generated ${rows.length} music clue pages.`);
};

export default generateMusicPages;
