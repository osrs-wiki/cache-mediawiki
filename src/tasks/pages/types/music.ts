import { writePageToFile } from "../pages.utils";

import { musicPageBuilder } from "@/mediawiki/pages/music";
import { dbRowToMusicTrack } from "@/mediawiki/templates/InfoboxMusic/InfoboxMusic.utils";
import { CacheProvider, DBRow } from "@/utils/cache2";

export const writeMusicPageFromCache = async (
  cache: Promise<CacheProvider>,
  id: number
) => {
  try {
    const dbRow = await DBRow.load(cache, id);

    if (dbRow && dbRow.table === 44) {
      writeMusicPage(dbRow);
    }
  } catch (e) {
    console.error(`Error generating page for music track ${id}: `, e);
  }
};

export const writeMusicPage = async (dbRow: DBRow) => {
  const musicTrack = dbRowToMusicTrack(dbRow);
  const trackName = musicTrack.displayName || musicTrack.sortName || `Track ${musicTrack.id}`;

  const builder = musicPageBuilder(musicTrack);
  writePageToFile(builder, "music", trackName, musicTrack.id.toString());
};