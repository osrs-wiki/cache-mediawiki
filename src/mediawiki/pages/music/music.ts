import {
  MediaWikiBreak,
  MediaWikiBuilder,
  MediaWikiTemplate,
  MediaWikiText,
} from "@osrs-wiki/mediawiki-builder";

import Context from "../../../context";

import InfoboxMusicTemplate from "@/mediawiki/templates/InfoboxMusic";
import { MusicTrack } from "@/types/music";

export const musicPageBuilder = (musicTrack: MusicTrack) => {
  const infoboxMusic = InfoboxMusicTemplate(musicTrack);

  const trackName =
    musicTrack.displayName || musicTrack.sortName || `Track ${musicTrack.id}`;

  const builder = new MediaWikiBuilder();

  if (Context.newContentTemplate) {
    builder.addContent(new MediaWikiTemplate(Context.newContentTemplate));
  }

  builder.addContents([
    infoboxMusic.build(),
    new MediaWikiText(trackName, { bold: true }),
    new MediaWikiText(" is a [[music]] track"),
  ]);

  // Add unlock location information if available in the hint
  if (musicTrack.unlockHint) {
    builder.addContents([
      new MediaWikiText(" that is unlocked "),
      new MediaWikiText(musicTrack.unlockHint),
    ]);
  } else {
    builder.addContent(new MediaWikiText("."));
  }

  builder.addContents([
    new MediaWikiBreak(),
    new MediaWikiBreak(),
    new MediaWikiTemplate("Music"),
    new MediaWikiText("[[Category:Old School-exclusive music]]"),
  ]);

  return builder;
};

export default musicPageBuilder;
