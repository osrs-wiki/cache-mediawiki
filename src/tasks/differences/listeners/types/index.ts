import { areaListener } from "./areas";
import { itemListener } from "./items";
import { musicListener } from "./music";
import { npcListener } from "./npcs";
import { objectListener } from "./objects";
import { CacheChangeListener } from "../listeners.types";

export const cacheListeners: CacheChangeListener[] = [
  areaListener,
  itemListener,
  musicListener,
  npcListener,
  objectListener,
];
