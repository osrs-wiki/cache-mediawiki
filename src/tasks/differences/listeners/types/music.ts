import { writeMusicPage } from "../../../pages/types";
import { CacheChangeListener } from "../listeners.types";

import Context from "@/context";
import {
  IndexType,
  ConfigType,
  DBRow,
  DBRowID,
  Reader,
} from "@/utils/cache2";

export const musicListener: CacheChangeListener = {
  index: IndexType.Configs,
  archive: ConfigType.DbRow,
  handler: async ({ oldFile, newFile }) => {
    const newEntry = newFile
      ? DBRow.decode(
          new Reader(newFile.file.data, {
            era: "osrs",
            indexRevision: newFile.index.revision,
          }),
          <DBRowID>newFile.file.id
        )
      : undefined;

    // Only process entries from table 44 (music tracks)
    if (
      Context.pages &&
      !oldFile &&
      newEntry &&
      newEntry.table === 44
    ) {
      writeMusicPage(newEntry);
    }
  },
};