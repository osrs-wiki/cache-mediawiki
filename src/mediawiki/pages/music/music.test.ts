import { musicPageBuilder } from "./music";

import { MusicTrack } from "@/types/music";

// Mock Context module
jest.mock("@/context", () => ({
  default: {
    updateDate: "2024-09-25",
    update: "Varlamore: The Rising Darkness is OUT NOW",
  },
}));

describe("musicPageBuilder", () => {
  const createMockMusicTrack = (
    sortName: string,
    displayName: string,
    unlockHint: string,
    duration: number,
    id = 123
  ): MusicTrack => {
    return {
      sortName,
      displayName,
      unlockHint,
      duration,
      id,
    };
  };

  it("should create a music page with proper structure", () => {
    const mockMusicTrack = createMockMusicTrack(
      "Stones of Old",
      "Stones of Old",
      "This track unlocks in Quetzacalli Gorge.",
      234
    );

    const builder = musicPageBuilder(mockMusicTrack);
    const result = builder.build();

    expect(result).toMatchSnapshot();
  });

  it("should handle quest detection properly", () => {
    const questTrack = createMockMusicTrack(
      "Quest Track",
      "Quest Track", 
      "This track unlocks during a quest.",
      180
    );

    const nonQuestTrack = createMockMusicTrack(
      "Regular Track",
      "Regular Track",
      "This track unlocks in an area.",
      120
    );

    const questBuilder = musicPageBuilder(questTrack);
    const questResult = questBuilder.build();

    const nonQuestBuilder = musicPageBuilder(nonQuestTrack);
    const nonQuestResult = nonQuestBuilder.build();

    // The quest detection happens in the InfoboxMusic template
    expect(questResult).toContain("quest = Yes");
    expect(nonQuestResult).toContain("quest = No");
  });

  it("should use display name over sort name", () => {
    const mockMusicTrack = createMockMusicTrack(
      "sort_name",
      "Display Name",
      "Test hint",
      150
    );

    const builder = musicPageBuilder(mockMusicTrack);
    const result = builder.build();

    expect(result).toContain("'''Display Name'''");
    expect(result).not.toContain("'''sort_name'''");
  });

  it("should fallback to sort name when display name is missing", () => {
    const mockMusicTrack = createMockMusicTrack(
      "Sort Name",
      "",
      "Test hint",
      150
    );

    const builder = musicPageBuilder(mockMusicTrack);
    const result = builder.build();

    expect(result).toContain("'''Sort Name'''");
  });
});