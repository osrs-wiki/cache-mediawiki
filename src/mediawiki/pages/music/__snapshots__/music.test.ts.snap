// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`musicPageBuilder should create a music page with proper structure 1`] = `
"{{Infobox Music
|name = Stones of Old
|file = [[File:Stones of Old.ogg]]
|members = Yes
|hint = This track unlocks in Quetzacalli Gorge.
|quest = No
|duration = 3:54
}}
'''Stones of Old''' is a [[music]] track that is unlocked in Quetzacalli Gorge.

{{Music}}
[[Category:Old School-exclusive music]]"
`;
